From bde71ab550fa3470adc63886f0cce1287d564ae5 Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:59:03 -0400
Subject: [scsi] Workaround for disks that report bad optimal transfer length

Message-id: <1408561176-18133-25-git-send-email-emilne@redhat.com>
Patchwork-id: 88458
O-Subject: [RHEL7.1 PATCH e-stor 24/57] [SCSI] Workaround for disks that report bad optimal transfer length
Bugzilla: 1132092
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit 56f2a8016e0ab54de8daaac3df4712cad0fcef2e

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit 56f2a8016e0ab54de8daaac3df4712cad0fcef2e
Author: Martin K. Petersen <martin.petersen@oracle.com>
Date:   Wed Apr 24 21:19:47 2013 -0400

    [SCSI] Workaround for disks that report bad optimal transfer length

    Not all disks fill out the VPD pages correctly. Add a blacklist flag
    that allows us ignore the SBC-3 VPD pages for a given device. The
    BLIST_SKIP_VPD_PAGES flag triggers our existing skip_vpd_pages
    scsi_device parameter to bypass VPD scanning.

    Also blacklist the offending Seagate drive model.

    Reported-by: Mike Snitzer <snitzer@redhat.com>
    Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/scsi_devinfo.c b/drivers/scsi/scsi_devinfo.c
index 43fca91..f969aca 100644
--- a/drivers/scsi/scsi_devinfo.c
+++ b/drivers/scsi/scsi_devinfo.c
@@ -228,6 +228,7 @@ static struct {
  {"SanDisk", "ImageMate CF-SD1", NULL, BLIST_FORCELUN},
  {"SEAGATE", "ST34555N", "0930", BLIST_NOTQ}, /* Chokes on tagged INQUIRY */
  {"SEAGATE", "ST3390N", "9546", BLIST_NOTQ},
+ {"SEAGATE", "ST900MM0006", NULL, BLIST_SKIP_VPD_PAGES},
  {"SGI", "RAID3", "*", BLIST_SPARSELUN},
  {"SGI", "RAID5", "*", BLIST_SPARSELUN},
  {"SGI", "TP9100", "*", BLIST_REPORTLUN2},
diff --git a/drivers/scsi/scsi_scan.c b/drivers/scsi/scsi_scan.c
index 852915a..307a811 100644
--- a/drivers/scsi/scsi_scan.c
+++ b/drivers/scsi/scsi_scan.c
@@ -926,6 +926,9 @@ static int scsi_add_lun(struct scsi_device *sdev, unsigned char *inq_result,
 
  sdev->eh_timeout = SCSI_DEFAULT_EH_TIMEOUT;
 
+ if (*bflags & BLIST_SKIP_VPD_PAGES)
+  sdev->skip_vpd_pages = 1;
+
  transport_configure_device(&sdev->sdev_gendev);
 
  if (sdev->host->hostt->slave_configure) {
diff --git a/include/scsi/scsi_devinfo.h b/include/scsi/scsi_devinfo.h
index cc1f3e7..447d2d7 100644
--- a/include/scsi/scsi_devinfo.h
+++ b/include/scsi/scsi_devinfo.h
@@ -31,4 +31,5 @@
 #define BLIST_MAX_512  0x800000 /* maximum 512 sector cdb length */
 #define BLIST_ATTACH_PQ3 0x1000000 /* Scan: Attach to PQ3 devices */
 #define BLIST_NO_DIF  0x2000000 /* Disable T10 PI (DIF) */
+#define BLIST_SKIP_VPD_PAGES 0x4000000 /* Ignore SBC-3 VPD pages */
 #endif
-- 
1.7.1